articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,248 @@
+[ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+However, that tutorial doesn't provide instructions to build other ROS 2
+packages, but just for building a whole ROS 2 distribution. Ideally, we should
+be able to cross-compile a ROS package for ROS 2 by launching a single
+`colcon build` command. Finally, cross-compiling for other platforms besides
+ARM64 is not supported in the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation
+tutorial, based on building a sysroot for the target platform using qemu and
+Docker, and then using `cmake-toolchains` with a toolchain file pointing
+`CMAKE_SYSROOT` to the sysroot location to compile with C and C++
+cross-compilers.
+
+On top of that we propose using a **colcon plugin `cc-build`** to wrap the
+instructions on the cross-compilation tutorial into a single build command.
+That plugin would define a new `colcon_core.verb` entry point pointing to a
+class `CCBuildVerb` that uses the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern)
+to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
+used by `colcon build` with additional functionality to handle the sysroot and
+toolchain file, and to inject additional `cmake-args` arguments. That would
+allows us to run the single command below from a ROS 2 overlay workspace on a
+developer workstation, for cross-compiling the packages in the workspace up to
+a package `performance_test` for the ARM-HF platform.
+
+```bash
+colcon cc-build --arch armhf --os ubuntu_bionic \
+  --packages-up-to performance_test # other `colcon build` arguments
+```
+
+That would should result on a new workspace directory `armhf-ubuntu_bionic`
+with subdirectories `build`, `install`, and `log` directories, with the
+ross-compiled binaries. All options for `colcon build` will be supported by
+forwarding them to the `BuildVerb` class. That would be the only command
+required for cross-compiling the packages in a workspace. Under the hood, the
+plugin would do the following:
+
+- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
+  (`qemu-user-static`, `g++-aarch64-linux-gnu`, ...) should be installed in the
+  development workstation used for the build. The `cc-build` plugin will
+  declare those system dependencies using a `stdeb.cfg` file like is done on
+  [other colcon plugins](https://github.com/colcon/colcon-core/blob/master/stdeb.cfg).
+- In order to build the sysroot, the plugin downloads a base ROS 2 Docker image
+  for the target platform, and builds a workspace dependent sysroot image by
+  doing a `COPY` to copy the contents of the workspace into the container, and
+  running `rosdep` on the copy. That ensures no workspace system dependency is
```

Comment:
I'm wondering whether we could execute a "well-known script" like `./initialize.sh` or something to avoid making the assumption that `rosdep` will always be the only required step there.

This would be helpful to test the plugin too as we could run non-ROS 2 images (e.g. base OS image only) to make sure we can cross-compile a simple "hello world" binary.

Reply:
That makes sense, but I'm not sure where to put it. Some options that come to my mind: 

- Make it dynamic: and optional argument for the plugin to provide and `./initialize.sh` to run at the end of the workspace image. If not provided we use a default `./initialize.sh` that runs rosdep. 
- Make it static: for each (os, arch) combination (or whichever definition of platform we use) the plugin uses a different Dockerfile,  extending a base Dockerfile that extends the base image for ROS 2 specified and does the COPY of the workspace contents. This means the plugin has to be modified to add or modify the Dockerfile specific for a platform. 

Does some of these option sound good to you?

Reply:
I would suggest holding off on this until it's necessary for a use case instead of speculatively extending the scope.